File: v4/rust/iprange-livedb/src/range_bulk.rs
//! Direct mapped-page construction of an ordered canonical range tree.

use std::marker::PhantomData;

use crate::contract::ValueKind;
use crate::error::{Error, Result};
use crate::fixed_tree::Store;
use crate::key::IpKey;
use crate::slotted_page::{Appender, PageSink};

const RANGE_BRANCH: u8 = 1;
const RANGE_LEAF: u8 = 2;
const BRANCH_LEVELS: usize = 6;
const MAX_RANGE_CELL: usize = 36;
const MAX_BRANCH_CELL: usize = 20;

pub(crate) trait Sink {
    type WritePage<'a>: PageSink
    where
        Self: 'a;

    fn allocate(&mut self) -> Result<u32>;
    fn update_page<'a, T, F>(&'a mut self, page_number: u32, update: F) -> Result<T>
    where
        F: FnOnce(&mut Self::WritePage<'a>) -> Result<T>;
}

impl<T: Store> Sink for T {
    type WritePage<'a>
        = T::WritePage<'a>
    where
        Self: 'a;

    fn allocate(&mut self) -> Result<u32> {
        Store::allocate(self)
    }

    fn update_page<'a, R, F>(&'a mut self, page_number: u32, update: F) -> Result<R>
    where
        F: FnOnce(&mut Self::WritePage<'a>) -> Result<R>,
    {
        Store::update_page(self, page_number, update)
    }
}

#[derive(Clone, Copy, Debug)]
pub(crate) struct Record<K> {
    pub(crate) from: K,
    pub(crate) to: K,
    pub(crate) value: u32,
}

#[derive(Clone, Copy, Debug)]
struct Node<K> {
    first: K,
    page_number: u32,
}

enum FinishedLevel<K> {
    Root(u32),
    Parent(Node<K>),
}

#[derive(Debug)]
struct PackedPage<K> {
    appender: Option<Appender>,
    first: Option<K>,
    page_number: Option<u32>,
}

impl<K: Copy> PackedPage<K> {
    const fn empty() -> Self {
        Self {
            appender: None,
            first: None,
            page_number: None,
        }
    }

    fn start<S: Sink>(
        &mut self,
        sink: &mut S,
        page_type: u8,
        born_txn: u64,
        level: u16,
        aux: u32,
    ) -> Result<()> {
        let page_number = sink.allocate()?;
        let appender = sink.update_page(page_number, |page| {
            Ok(Appender::new(page, page_type, born_txn, level, aux))
        })?;
        self.appender = Some(appender);
        self.first = None;
        self.page_number = Some(page_number);
        Ok(())
    }

    fn push<S: Sink>(&mut self, sink: &mut S, first: K, cell: &[u8]) -> Result<bool> {
        let page_number = self
            .page_number
            .ok_or(Error::Corrupt("ordered range page has no output page"))?;
        let appender = self
            .appender
            .as_mut()
            .ok_or(Error::Corrupt("ordered range page is not active"))?;
        let appended = sink.update_page(page_number, |page| appender.try_push(page, cell))?;
        if appended && self.first.is_none() {
            self.first = Some(first);
        }
        Ok(appended)
    }

    fn finish<S: Sink>(&mut self, sink: &mut S) -> Result<Node<K>> {
        let appender = self
            .appender
            .take()
            .ok_or(Error::Corrupt("ordered range page is not active"))?;
        let page_number = self
            .page_number
            .take()
            .ok_or(Error::Corrupt("ordered range page has no output page"))?;
        sink.update_page(page_number, |page| appender.finish(page))?;
        let first = self
            .first
            .take()
            .ok_or(Error::Corrupt("ordered range page has no first key"))?;
        Ok(Node { first, page_number })
    }

    fn active(&self) -> bool {
        self.appender.is_some()
    }
}

#[derive(Debug)]
struct BranchLevel<K> {
    page: PackedPage<K>,
    only_child: Option<Node<K>>,
    emitted: bool,
}

impl<K: Copy> BranchLevel<K> {
    const fn empty() -> Self {
        Self {
            page: PackedPage::empty(),
            only_child: None,
            emitted: false,
        }
    }
}

#[derive(Debug)]
pub(crate) struct Builder<K> {
    born_txn: u64,
    value_kind: ValueKind,
    leaf: PackedPage<K>,
    branches: [BranchLevel<K>; BRANCH_LEVELS],
    previous: Option<Record<K>>,
    record_count: u64,
    key: PhantomData<K>,
}

impl<K: IpKey> Builder<K> {
    pub(crate) fn new(born_txn: u64, value_kind: ValueKind) -> Self {
        Self {
            born_txn,
            value_kind,
            leaf: PackedPage::empty(),
            branches: std::array::from_fn(|_| BranchLevel::empty()),
            previous: None,
            record_count: 0,
            key: PhantomData,
        }
    }

    pub(crate) fn push<S: Sink>(&mut self, sink: &mut S, record: Record<K>) -> Result<()> {
        self.require_record(record)?;
        let next_count = self
            .record_count
            .checked_add(1)
            .ok_or(Error::ArithmeticOverflow("range record count"))?;
        let mut cell = [0; MAX_RANGE_CELL];
        let cell_len = K::WIDTH * 2 + 4;
        record.from.write_le(&mut cell);
        record.to.write_le(&mut cell[K::WIDTH..]);
        cell[K::WIDTH * 2..cell_len].copy_from_slice(&record.value.to_le_bytes());
        self.push_leaf_cell(sink, record.from, &cell[..cell_len])?;
        self.previous = Some(record);
        self.record_count = next_count;
        Ok(())
    }

    fn push_leaf_cell<S: Sink>(&mut self, sink: &mut S, first: K, cell: &[u8]) -> Result<()> {
        if !self.leaf.active() {
            self.leaf
                .start(sink, RANGE_LEAF, self.born_txn, 0, K::FAMILY as u32)?;
        }
        if self.leaf.push(sink, first, cell)? {
            return Ok(());
        }
        let node = self.leaf.finish(sink)?;
        self.push_node(sink, 0, node)?;
        self.leaf
            .start(sink, RANGE_LEAF, self.born_txn, 0, K::FAMILY as u32)?;
        if self.leaf.push(sink, first, cell)? {
            Ok(())
        } else {
            Err(Error::Corrupt("range record does not fit an empty leaf"))
        }
    }

    fn require_record(&self, record: Record<K>) -> Result<()> {
        if record.from > record.to {
            return Err(Error::InvalidArgument("range start is after its end"));
        }
        if self.value_kind == ValueKind::Membership && record.value == 0 {
            return Err(Error::Corrupt("membership range value is zero"));
        }
        let Some(previous) = self.previous else {
            return Ok(());
        };
        if previous.to >= record.from {
            return Err(Error::InvalidArgument(
                "ordered output ranges are not strictly increasing",
            ));
        }
        if previous.value == record.value && previous.to.checked_next() == Some(record.from) {
            return Err(Error::InvalidArgument(
                "adjacent equal ranges are not canonical",
            ));
        }
        Ok(())
    }

    fn push_node<S: Sink>(
        &mut self,
        sink: &mut S,
        level_index: usize,
        node: Node<K>,
    ) -> Result<()> {
        if level_index == BRANCH_LEVELS {
            return Err(Error::PageSpaceExhausted);
        }

        if !self.branches[level_index].page.active() {
            if let Some(first) = self.branches[level_index].only_child.take() {
                self.start_branch(sink, level_index)?;
                if !self.push_branch_cell(sink, level_index, first)? {
                    return Err(Error::Corrupt("range branch cell does not fit"));
                }
            } else {
                self.branches[level_index].only_child = Some(node);
                return Ok(());
            }
        }
        if self.push_branch_cell(sink, level_index, node)? {
            return Ok(());
        }

        let parent = self.branches[level_index].page.finish(sink)?;
        self.branches[level_index].emitted = true;
        self.push_node(sink, level_index + 1, parent)?;
        self.branches[level_index].only_child = Some(node);
        Ok(())
    }

    fn start_branch<S: Sink>(&mut self, sink: &mut S, level_index: usize) -> Result<()> {
        self.branches[level_index].page.start(
            sink,
            RANGE_BRANCH,
            self.born_txn,
            level_index as u16 + 1,
            K::FAMILY as u32,
        )
    }

    fn push_branch_cell<S: Sink>(
        &mut self,
        sink: &mut S,
        level_index: usize,
        node: Node<K>,
    ) -> Result<bool> {
        let mut cell = [0; MAX_BRANCH_CELL];
        let cell_len = K::WIDTH + 4;
        node.first.write_le(&mut cell);
        cell[K::WIDTH..cell_len].copy_from_slice(&node.page_number.to_le_bytes());
        self.branches[level_index]
            .page
            .push(sink, node.first, &cell[..cell_len])
    }

    pub(crate) fn finish<S: Sink>(&mut self, sink: &mut S) -> Result<(u32, u64)> {
        if self.record_count == 0 {
            return Ok((0, 0));
        }
        let leaf = self.leaf.finish(sink)?;
        self.push_node(sink, 0, leaf)?;

        for level_index in 0..BRANCH_LEVELS {
            match self.finish_level(sink, level_index)? {
                None => {}
                Some(FinishedLevel::Root(root)) => return Ok((root, self.record_count)),
                Some(FinishedLevel::Parent(node)) => {
                    self.push_node(sink, level_index + 1, node)?;
                }
            }
        }
        Err(Error::PageSpaceExhausted)
    }

    fn finish_level<S: Sink>(
        &mut self,
        sink: &mut S,
        level_index: usize,
    ) -> Result<Option<FinishedLevel<K>>> {
        if self.branches[level_index].page.active() {
            let node = self.branches[level_index].page.finish(sink)?;
            return Ok(Some(if self.branches[level_index].emitted {
                FinishedLevel::Parent(node)
            } else {
                FinishedLevel::Root(node.page_number)
            }));
        }
        let Some(child) = self.branches[level_index].only_child.take() else {
            return Ok(None);
        };
        if !self.branches[level_index].emitted {
            return Ok(Some(FinishedLevel::Root(child.page_number)));
        }
        self.start_branch(sink, level_index)?;
        if !self.push_branch_cell(sink, level_index, child)? {
            return Err(Error::Corrupt("range branch cell does not fit"));
        }
        let node = self.branches[level_index].page.finish(sink)?;
        Ok(Some(FinishedLevel::Parent(node)))
    }
}

File: v4/go/internal/writer/range_bulk.go
// Direct mapped-page construction of an ordered canonical range tree
// (Rust range_bulk.rs): a 6-level bottom-up builder with only-child
// collapse. Range records append in canonical order; full leaves roll
// upward as branch pages; finish yields the root and the record count.

package writer

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
	"github.com/firehol/iprange/v4/go/internal/work"
)

const rangeBulkBranchLevels = 6

// rangeBulkNode is one finished page plus its first key (Rust Node).
type rangeBulkNode[K any] struct {
	first      K
	pageNumber uint32
}

// rangeBulkPackedPage is one page under construction (Rust PackedPage).
type rangeBulkPackedPage[K any] struct {
	appender   format.SlottedAppender
	started    bool
	first      K
	hasFirst   bool
	pageNumber uint32
}

func (p *rangeBulkPackedPage[K]) active() bool { return p.started }

func (p *rangeBulkPackedPage[K]) start(store tree.Store, pageType format.PageType, bornTxn uint64, level uint16, aux uint32) error {
	pageNumber, err := store.Allocate()
	if err != nil {
		return err
	}
	page, tag, err := store.Update(pageNumber)
	if err != nil {
		return err
	}
	appender := format.NewSlottedAppender(page, pageType, bornTxn, level, aux)
	if err := store.RestoreDirty(pageNumber, tag); err != nil {
		return err
	}
	p.appender = appender
	p.started = true
	var zero K
	p.first = zero
	p.hasFirst = false
	p.pageNumber = pageNumber
	return nil
}

func (p *rangeBulkPackedPage[K]) push(store tree.Store, first K, cell []byte) (bool, error) {
	if p.pageNumber == 0 {
		return false, corrupt("ordered range page has no output page")
	}
	if !p.started {
		return false, corrupt("ordered range page is not active")
	}
	page, tag, err := store.Update(p.pageNumber)
	if err != nil {
		return false, err
	}
	appended, err := p.appender.TryPush(page, cell)
	if err != nil {
		return false, err
	}
	if err := store.RestoreDirty(p.pageNumber, tag); err != nil {
		return false, err
	}
	if appended && !p.hasFirst {
		p.first = first
		p.hasFirst = true
	}
	return appended, nil
}

func (p *rangeBulkPackedPage[K]) finish(store tree.Store) (rangeBulkNode[K], error) {
	appender := &p.appender
	if !p.started {
		return rangeBulkNode[K]{}, corrupt("ordered range page is not active")
	}
	pageNumber := p.pageNumber
	if pageNumber == 0 {
		return rangeBulkNode[K]{}, corrupt("ordered range page has no output page")
	}
	page, tag, err := store.Update(pageNumber)
	if err != nil {
		return rangeBulkNode[K]{}, err
	}
	if err := appender.Finish(page); err != nil {
		return rangeBulkNode[K]{}, err
	}
	if err := store.RestoreDirty(pageNumber, tag); err != nil {
		return rangeBulkNode[K]{}, err
	}
	if !p.hasFirst {
		return rangeBulkNode[K]{}, corrupt("ordered range page has no first key")
	}
	first := p.first
	p.started = false
	var zero K
	p.first = zero
	p.hasFirst = false
	p.pageNumber = 0
	return rangeBulkNode[K]{first: first, pageNumber: pageNumber}, nil
}

type rangeBulkBranchLevel[K any] struct {
	page      rangeBulkPackedPage[K]
	onlyChild rangeBulkNode[K]
	hasOnly   bool
	emitted   bool
}

type rangeBulkBuilder[K any] struct {
	bornTxn     uint64
	valueKind   uint8
	leaf        rangeBulkPackedPage[K]
	branches    [rangeBulkBranchLevels]rangeBulkBranchLevel[K]
	previous    rangeRecord[K]
	hasPrevious bool
	recordCount uint64
	family      rangeFamily[K]
	// recordScratch and branchScratch own the encoded record and branch
	// cells of one push (Rust Builder local cells). The codec interface
	// methods retain their slice argument, so a stack cell would escape
	// per record; the builder-owned buffers make the record push
	// allocation-free exactly like the Rust locals.
	recordScratch [rangeRecordMaxSize]byte
	branchScratch [rangeBranchMaxSize]byte
}

// newRangeBulkBuilder starts one ordered range tree builder (Rust
// Builder::new). family selects the address family of the records.
func newRangeBulkBuilder[K any](bornTxn uint64, valueKind uint8, family rangeFamily[K]) *rangeBulkBuilder[K] {
	return &rangeBulkBuilder[K]{bornTxn: bornTxn, valueKind: valueKind, family: family}
}

// init starts one ordered range tree builder in place (Rust
// Builder::new): the coverage ordered prefix embeds its builder so the
// per-workflow ordered path never allocates.
func (b *rangeBulkBuilder[K]) init(bornTxn uint64, valueKind uint8, family rangeFamily[K]) {
	*b = rangeBulkBuilder[K]{bornTxn: bornTxn, valueKind: valueKind, family: family}
}

var errRangeNotCanonical = invalid("ordered output ranges are not canonical")

// push appends one canonical range record; out-of-order or adjacent
// same-value records are rejected (Rust Builder::push).
func (b *rangeBulkBuilder[K]) push(store tree.Store, record rangeRecord[K]) error {
	ok, err := b.tryPush(store, record)
	if err != nil {
		return err
	}
	if !ok {
		return errRangeNotCanonical
	}
	return nil
}

// tryPush appends the record when it is canonical and reports whether it
// was appended (Rust Builder::try_push).
func (b *rangeBulkBuilder[K]) tryPush(store tree.Store, record rangeRecord[K]) (bool, error) {
	ok, err := b.canAppend(record)
	if err != nil || !ok {
		return ok, err
	}
	b.recordCount++
	cell := b.recordScratch[:]
	cellLen, err := b.family.EncodeRecord(record, cell)
	if err != nil {
		return false, err
	}
	if err := b.pushLeafCell(store, record.from, cell[:cellLen]); err != nil {
		return false, err
	}
	b.previous = record
	b.hasPrevious = true
	work.RangeEmitted(1)
	return true, nil
}

func (b *rangeBulkBuilder[K]) pushLeafCell(store tree.Store, first K, cell []byte) error {
	if !b.leaf.active() {
		if err := b.leaf.start(store, format.PageTypeRangeLeaf, b.bornTxn, 0, rangeFamilyAux(b.family)); err != nil {
			return err
		}
	}
	if pushed, err := b.leaf.push(store, first, cell); err != nil {
		return err
	} else if pushed {
		return nil
	}
	node, err := b.leaf.finish(store)
	if err != nil {
		return err
	}
	if err := b.pushNode(store, 0, node); err != nil {
		return err
	}
	if err := b.leaf.start(store, format.PageTypeRangeLeaf, b.bornTxn, 0, rangeFamilyAux(b.family)); err != nil {
		return err
	}
	if pushed, err := b.leaf.push(store, first, cell); err != nil {
		return err
	} else if !pushed {
		return corrupt("range record does not fit an empty leaf")
	}
	return nil
}

func (b *rangeBulkBuilder[K]) canAppend(record rangeRecord[K]) (bool, error) {
	if b.family.Less(record.to, record.from) {
		return false, invalid("range start is after its end")
	}
	if b.valueKind != format.ValueKindDirect && record.value == 0 {
		return false, corrupt("indirect range value is zero")
	}
	if !b.hasPrevious {
		return true, nil
	}
	previous := b.previous
	if b.family.Equal(previous.to, record.from) || !b.family.Less(previous.to, record.from) {
		return false, nil
	}
	// The comparator above covers from<=to; the canonical rules reject
	// overlap and adjacency with the same value (Rust can_append): from
	// must be strictly after the previous to, and adjacent same-value
	// ranges must be merged by the caller.
	next, ok := b.family.Next(previous.to)
	if ok && previous.value == record.value && b.family.Equal(next, record.from) {
		return false, nil
	}
	return true, nil
}

func (b *rangeBulkBuilder[K]) pushNode(store tree.Store, levelIndex int, node rangeBulkNode[K]) error {
	if levelIndex == rangeBulkBranchLevels {
		return pageSpaceExhausted()
	}
	if !b.branches[levelIndex].page.active() {
		if b.branches[levelIndex].hasOnly {
			first := b.branches[levelIndex].onlyChild
			b.branches[levelIndex].hasOnly = false
			if err := b.startBranch(store, levelIndex); err != nil {
				return err
			}
			if pushed, err := b.pushBranchCell(store, levelIndex, first); err != nil {
				return err
			} else if !pushed {
				return corrupt("range branch cell does not fit")
			}
		} else {
			b.branches[levelIndex].onlyChild = node
			b.branches[levelIndex].hasOnly = true
			return nil
		}
	}
	if pushed, err := b.pushBranchCell(store, levelIndex, node); err != nil {
		return err
	} else if pushed {
		return nil
	}
	parent, err := b.branches[levelIndex].page.finish(store)
	if err != nil {
		return err
	}
	b.branches[levelIndex].emitted = true
	if err := b.pushNode(store, levelIndex+1, parent); err != nil {
		return err
	}
	b.branches[levelIndex].onlyChild = node
	b.branches[levelIndex].hasOnly = true
	return nil
}

func (b *rangeBulkBuilder[K]) startBranch(store tree.Store, levelIndex int) error {
	aux := rangeFamilyAux(b.family)
	return b.branches[levelIndex].page.start(store, format.PageTypeRangeBranch, b.bornTxn, uint16(levelIndex)+1, aux)
}

func (b *rangeBulkBuilder[K]) pushBranchCell(store tree.Store, levelIndex int, node rangeBulkNode[K]) (bool, error) {
	cell := b.branchScratch[:]
	cellLen, err := encodeRangeBranch(b.family, node.first, node.pageNumber, cell)
	if err != nil {
		return false, err
	}
	return b.branches[levelIndex].page.push(store, node.first, cell[:cellLen])
}

// finish seals the tree and returns the root and the record count (Rust
// Builder::finish; the output-pass charge separates whole-tree builds
// from the coverage ordered-prefix builds).
func (b *rangeBulkBuilder[K]) finish(store tree.Store) (uint32, uint64, error) {
	work.OutputPass(1)
	return b.finishInline(store)
}

// finishInline seals the tree without charging a whole output pass (Rust
// Builder::finish_inline; the coverage ordered prefix seals inline inside
// one input workflow, so the pass is charged by the merge that consumes
// it, never here).
func (b *rangeBulkBuilder[K]) finishInline(store tree.Store) (uint32, uint64, error) {
	if b.recordCount == 0 {
		return 0, 0, nil
	}
	leaf, err := b.leaf.finish(store)
	if err != nil {
		return 0, 0, err
	}
	if err := b.pushNode(store, 0, leaf); err != nil {
		return 0, 0, err
	}
	for levelIndex := 0; levelIndex < rangeBulkBranchLevels; levelIndex++ {
		finished, err := b.finishLevel(store, levelIndex)
		if err != nil {
			return 0, 0, err
		}
		switch f := finished.(type) {
		case nil:
		case rangeBulkFinishedRoot:
			return uint32(f), b.recordCount, nil
		case rangeBulkFinishedParent[K]:
			if err := b.pushNode(store, levelIndex+1, rangeBulkNode[K](f)); err != nil {
				return 0, 0, err
			}
		}
	}
	return 0, 0, pageSpaceExhausted()
}

type rangeBulkFinishedRoot uint32
type rangeBulkFinishedParent[K any] rangeBulkNode[K]

func (b *rangeBulkBuilder[K]) finishLevel(store tree.Store, levelIndex int) (any, error) {
	if b.branches[levelIndex].page.active() {
		node, err := b.branches[levelIndex].page.finish(store)
		if err != nil {
			return nil, err
		}
		if b.branches[levelIndex].emitted {
			return rangeBulkFinishedParent[K](node), nil
		}
		return rangeBulkFinishedRoot(node.pageNumber), nil
	}
	if !b.branches[levelIndex].hasOnly {
		return nil, nil
	}
	child := b.branches[levelIndex].onlyChild
	b.branches[levelIndex].hasOnly = false
	if !b.branches[levelIndex].emitted {
		return rangeBulkFinishedRoot(child.pageNumber), nil
	}
	if err := b.startBranch(store, levelIndex); err != nil {
		return nil, err
	}
	if pushed, err := b.pushBranchCell(store, levelIndex, child); err != nil {
		return nil, err
	} else if !pushed {
		return nil, corrupt("range branch cell does not fit")
	}
	node, err := b.branches[levelIndex].page.finish(store)
	if err != nil {
		return nil, err
	}
	return rangeBulkFinishedParent[K](node), nil
}

func pageSpaceExhausted() error {
	return &format.Error{Code: format.CodePageSpaceExhausted, Detail: "v4 page-number space is exhausted"}
}

// rangeFamilyAux reports the mapped page AUX value of one range family
// codec (the format address family byte).
func rangeFamilyAux[K any](codec rangeFamily[K]) uint32 {
	if _, ok := any(codec).(rangeCodec4); ok {
		return uint32(format.AddressFamilyIPv4)
	}
	return uint32(format.AddressFamilyIPv6)
}

const (
	rangeRecordMaxSize = 36
	rangeBranchMaxSize = 20
)

// encodeRangeBranch writes one family branch cell (key prefix plus the
// child page number; Rust RangeCodec::write_branch).
func encodeRangeBranch[K any](codec rangeFamily[K], first K, child uint32, output []byte) (int, error) {
	size := codec.KeySize()
	codec.WriteKey(codec.KeyOf(first), output[0:size])
	format.PutU32(output[size:], child)
	return size + 4, nil
}
